objects/src/accounts/vault.rs

```diff
@@ -59,11 +70,169 @@ impl AccountVault {
         if !matches!(faucet_id.account_type(), AccountType::FungibleFaucet) {
             return Err(AccountError::not_a_fungible_faucet_id(faucet_id));
         }
-        todo!()
+
+        // if the tree value is [0, 0, 0, 0], the asset is not stored in the vault
+        match self.asset_tree.get_value([ZERO, ZERO, ZERO, faucet_id.into()].into()) {
+            asset if asset == [ZERO, ZERO, ZERO, ZERO] => Ok(0),
+            asset => Ok(FungibleAsset::try_from(asset)
+                .expect("tree only contains valid assets")
+                .amount()),
+        }
     }
 
     /// Returns a list of assets stored in this vault.
     pub fn assets(&self) -> &[Asset] {
         &self.assets
     }
+
+    // PUBLIC MODIFIERS
+    // --------------------------------------------------------------------------------------------
+
+    // ADD ASSET
+    // --------------------------------------------------------------------------------------------
+    /// Add the specified asset to the vault.
+    pub fn add_asset(&mut self, asset: Asset) -> Result<Asset, AccountError> {
+        Ok(match asset {
+            Asset::Fungible(asset) => Asset::Fungible(self.add_fungible_asset(asset)?),
+            Asset::NonFungible(asset) => Asset::NonFungible(self.add_non_fungible_asset(asset)?),
+        })
+    }
+
+    /// Add the specified fungible asset to the vault.  If the vault already contains an asset
+    /// issued by the same faucet, the amounts are added together.
+    ///
+    /// # Errors
+    /// - If the total value of assets is greater than or equal to 2^63.
+    fn add_fungible_asset(&mut self, asset: FungibleAsset) -> Result<FungibleAsset, AccountError> {
+        // fetch current asset value from the tree and add the new asset to it.
+        let new: FungibleAsset = match self.asset_tree.get_value(asset.vault_key().into()) {
+            current if current == [ZERO, ZERO, ZERO, ZERO] => {
+                self.assets.push(Asset::Fungible(asset));
+                asset
+            }
+            _ => self
+                .assets
+                .iter_mut()
+                .find_map(|current| match current {
+                    Asset::Fungible(current_inner) if current_inner.is_from_same_faucet(&asset) => {
+                        let new = current_inner
+                            .add(asset)
+                            .map_err(AccountError::ModifyFungibleAssetBalanceError)
```

Comment:
I think the check for whether two fungible assets can be added is handled in `FungibleAsset::add()` - so, it may not be necessary here.